Allow list of common chunks as response to `match' request

joaotavora opened the pull request:
Also, in the company-capf backend, attempt to call :company-match in case
capf definition provides that.

* company-capf.el (company-capf): Look for :company-match prop
in the capf-function return value and use it for `match'.

* company.el (company-backends): Describe new meaning of `match'.o
(company-fill-propertize): Generalize to listp match.

My use case is my other package SLY (https:/github.com/joaotavora/sly) and its ability to "flex-search" completions.  The animated gif should illustrate this:

![company-flex-completion](https://user-images.githubusercontent.com/387011/41186054-9eef5fe4-6b87-11e8-8734-e6abb6a356b8.gif)

dgutov commented:
Thank you! I'll try to review it soon.

When used with previous versions of Emacs, it at least highlighted the part until the "first difference". But some update broke that. :-( And I hadn't had the time to fix it.

joaotavora commented:
> When used with previous versions of Emacs, it at least highlighted the part until the "first difference". But some update broke that. :-( And I hadn't had the time to fix it.

But do you mean `match` in general or the way `company-capf` provides match?

dgutov commented:
`company-capf`, yes.

joaotavora commented:
OK. I can have a look and maybe fix that...

dgutov commented:
Speaking of `match`, BTW. We've had a recent discussion in #762.

As far as I'm concerned, the takeaways are:

* Rename `match` to `highlights`.
* Allow it to return a list of integers.

Does returning a list of pairs make your life easier somehow?

joaotavora commented:
I'll have a look.  But what are the semantics of the list of integers?

joaotavora commented:
It seems to be working here...

![peek 2018-06-15 23-30](https://user-images.githubusercontent.com/387011/41491995-2f089f34-70f4-11e8-9187-7ae015df4f41.gif)

dgutov commented:
All screenshotting software failed me today :(

Try typing `(d-a-t`. It should highlight either `define-abbrev-t` (like it did before), or `d`, `a` and `t` separately. It only highlights `d`.

joaotavora commented:
> Try typing (d-a-t. It should highlight either define-abbrev-t (like it did before), or d, a and t separately. It only highlights d.

Nice, I see it. I'll have a look.

joaotavora commented:
> > Try typing (d-a-t. It should highlight either define-abbrev-t (like it did before), or d, a and t separately. It only highlights d.

> Nice, I see it. I'll have a look.

Hi @dgutov, so I did this and updated the PR. I hope I didn't trip any wires, so have a look.  In particular, I was a bit confused by the mention of `base-size` that was in `company-capf.el` before, but I don't think it applies to this new strategy of sniffing for the `completions-common-part` face.

 Anyway, here are some gifs.

![company-match-chunks](https://user-images.githubusercontent.com/387011/41983032-dbbca8fe-7a24-11e8-810b-dcf5e931632b.gif)
![company-define-abbrev-table](https://user-images.githubusercontent.com/387011/41983042-df6a5514-7a24-11e8-865b-e93eaecb5591.gif)

joaotavora commented:
BTW, I don't see any reason why this shouldn't fix #762, but maybe I'm missing something from that discussion.

dgutov commented on company-capf.el:
```diff
@@ -112,18 +112,27 @@
+           (while (< pos limit)
+             (setq pos
+                   (if (< pos 0) 0 (next-property-change pos arg limit)))
+             (setq prop-value (or (get-text-property pos 'face arg)
```
Do we really need the `or` here? Since `font-lock-face` is an alias for `face` in `char-property-alias-alist`, the fist clause should be sufficient.

dgutov commented on company-capf.el:
```diff
@@ -112,18 +112,27 @@
-            (text-property-not-all start (length arg)
-                                   'face value arg))
-          (length arg)))))
+     ;; ask the for the `:company-match' function.  If that doesn't help,
```
Please capitalize and punctuate the comments, even if existing ones don't always follow that convention.

dgutov commented on test/capf-tests.el:
```diff
@@ -0,0 +1,115 @@
+;;; capf-tests.el --- company tests for the company-capf backend  -*- lexical-binding: t; -*-
+
+;; Copyright (C) 2013-2015  Free Software Foundation, Inc.
```
The years are wrong here.

joaotavora commented:
All fixed.

dgutov commented:
Thank you very much. Let's deal with the high-level concerns now, and I'll merge.

> I don't see any reason why this shouldn't fix #762, but maybe I'm missing something from that discussion.

I mentioned the "missing" items, and you have not commented on them thus far.

* I do think it's better to rename the action to `highlights`, even if internally they will be aliases. At least in the `company-backends` docstring. Also, `company-highlights` capf property has a better ring to it, don't you think?

* I *think* a list of indexes, rather than an alist, is a better fit semantically (when we're doing fuzzy matching, we match all characters individually, after all). And it's a tad easier to produce (at least, using Elisp). I'm not sure if calling `font-lock-prepend-text-property` for each character is going to make a difference performance-wise, but if it does, we could glue the highlights together on our side.

On the other hand, if highlights were a simple list, you can't do `(if (integerp common) `((0 . ,common)) common)` so easily. And your default implementation in `company-capf` is also more amenable to returning a list of pairs. So... I'm on the fence here.

dgutov commented on company.el:
```diff
@@ -2515,11 +2517,17 @@ If SHOW-VERSION is non-nil, show the version in the echo area."
+     for inline-beg = (+ margin comp-beg)
+     for inline-end = (min (+ margin comp-end) width)
+     when (< inline-beg width)
+     do (font-lock-prepend-text-property inline-beg inline-end 'face
```
Should we also have a test in `frontends-tests` that makes sure that the complex case here renders successfully?

joaotavora commented:
> I do think it's better to rename the action to highlights, even if internally they will be aliases. At least in the company-backends docstring. Also, company-highlights capf property has a better ring to it, don't you think?

This is entirely cosmetic right, i.e. there's no extra functionality in `highlights`, right? If so, who am I to say what rings better :-)? (highlights does ring better tho). But do you want me to make this change too?

> I think a list of indexes, rather than an alist, is a better fit semantically (when we're doing fuzzy matching, we match all characters individually, after all). And it's a tad easier to produce (at least, using Elisp). I'm not sure if calling font-lock-prepend-text-property for each character is going to make a difference performance-wise, but if it does, we could glue the highlights together on our side. On the other hand, if highlights were a simple list, you can't do (if (integerp common) `((0 . ,common)) common) so easily. And your default implementation in company-capf is also more amenable to returning a list of pairs. So... I'm on the fence here

I don't agree that it's a better fit semantically at all.  Of course everything boils down to characters, but some engines might match by regexp, for example, who knows.  And generally regexp matching groups don't describe themselves by the character, they use regions/spans/chunks, right?  Why force these backends to convert to single char indexes?

Secondly, I think I noted before that the region/chunk approach is more economic memory-wise (which doesn't sound very relevant for "local" completions, but may be in network-connected backends such as SLY, SLIME, (future LSP?) when sending a big bunch of candidates across a slow connection).

But we don't have to agree on this at all if we use the following semantics:

* if a integer `X` treat it as the single region `((0 . X))`;
* if a list of integers `(X1 X2 .. Xn)`, treat as `n` regions `((X1 . X1 +1) (X2 . X2+1) (Xn . Xn+1))` (or is it rather `-1` and i'm off by one? you get the point);
* if a list of regions, treat as a list of regions.

Do you want me to do this?

Sarcasm commented:
Wow, thank you @joaotavora for tackling this.
The flymake improvements, the eglot client, and this contribution, it's very much appreciated!

With your patch, and tweaking a bit a company backend, I could the following output, which is exactly what I was after in #762:

![image](https://user-images.githubusercontent.com/424046/42112414-08d5b1cc-7be8-11e8-90d2-855060d1901f.png)

I don't really have comments on the API:
- single characters or ranges, both make sense to me
- `match` or `highlights`, if no API existed, I guess I would use `highlights`, but if `match` can be extended, why not

joaotavora commented:
> Wow, thank you @joaotavora for tackling this.

@Sarcasm you're welcome. I also agree with your judgement of the 2 outstanding bits.

@Dmitry, btw when removing that `or` I had to keep the second (`font-lock-face`), not the first clause.

dgutov commented:
You guys make very good points. :slightly_smiling_face: 

Regarding this, though:

> but some engines might match by regexp, for example, who knows. And generally regexp matching groups don't describe themselves by the character, they use regions/spans/chunks, right? Why force these backends to convert to single char indexes?

I wrote an example Elisp function that performs this kind of matching using an auto-generated regexp, here: https://github.com/company-mode/company-mode/issues/762#issuecomment-370188173

There's no obvious way to change the regexp to match spans of characters, is there? Please feel free to prove me wrong, or point out a different approach used somewhere else.

Though this is not essential, I guess, since we can collate consecutive numbers into spans "by hand" anyway.

dgutov commented:
> btw when removing that or I had to keep the second (font-lock-face), not the first clause

That's odd, I tried `(get-text-property 1 'face (buffer-substring (1- (point)) (1+ (point))))` interactively, in buffers where `face` is used (`emacs-lisp-mode`), as well as where `font-lock-face` is used (`js2-mode`), and it returned non-nil both times. Is this code somehow evaluated in a context where `char-property-alias-alist` is nil?

dgutov commented:
> But we don't have to agree on this at all if we use the following semantics:

I'd rather not overload the return type too much, unless we really find the cases for each of the return types where it gives a distinct advantage.

joaotavora commented:
> There's no obvious way to change the regexp to match spans of characters, is there? Please feel free to prove me wrong, or point out a different approach used somewhere else.

Perhaps I'm missing something, but what I meant is that, in general, regexp match groups report start and end positions, which I called "spans". You seem to have used a particular regexp where these spans measure 1 in length, but in general they measure 1 or more.  So using regions is, in general, a better match for regexp-based backends in the wild.

> That's odd, I tried (get-text-property 1 'face (buffer-substring (1- (point)) (1+ (point)))) interactively, in buffers where face is used (emacs-lisp-mode), as well as where font-lock-face is used (js2-mode), and it returned non-nil both times. Is this code somehow evaluated in a context where char-property-alias-alist is nil?

No idea, but I don't think so.  I just noticed how it failed the tests when I used your suggestion.  and I ran tests with `make test-batch`.

> I'd rather not overload the return type too much, unless we really find the cases for each of the return types where it gives a distinct advantage.

I was just catering to your specific request.  I personally can't find any advantage to the single-character index case, but that doesn't mean it doesn't exist. It depends on whether there are backends that produce highlighting information closer to that format.

Regarding the overloading, it seems perfectly progressive to me.  I bet the implementation for that extra bit is at most a line longer.

Since this seems to solve @Sarcasm's and my use case (and I think I've demonstrated that the single-character case can be represented suitably with regions), is there anything holding up this PR?

dgutov commented:
> You seem to have used a particular regexp where these spans measure 1 in length, but in general they measure 1 or more.

To reach the goal of supporting fuzzy matching, as it's normally understood these days (meaning every character can be anywhere). So I'm asking for examples where this is not so. But I suppose the "partial completion" completion style is one such example.

> No idea, but I don't think so. I just noticed how it failed the tests when I used your suggestion. and I ran tests with make test-batch.

That's probably because the batch mode doesn't turn on font-lock-mode. Oh well.

> is there anything holding up this PR?

Not much. Could you add a NEWS entry?

joaotavora commented:
> So I'm asking for examples where this is not so. But I suppose the "partial completion" completion style is one such example.

So is ido.el's default matching mechanism.

> That's probably because the batch mode doesn't turn on font-lock-mode. Oh well.

Oh good, right? i.e. looking for `face` is also correct, right?

> Not much. Could you add a NEWS entry?

Done.

dgutov commented on NEWS.md:
```diff
@@ -2,6 +2,14 @@
+  ([#798](https://github.com/company-mode/company-mode/issues/798),
+  [#762](https://github.com/company-mode/company-mode/issues/762))
+* The `company-capf` backend will pick up on a `:company-match` metadata element
+  on the capf function (similar `:company-location` or `:company-doc-buffer`)
```
"similar to"

dgutov commented on company-capf.el:
```diff
@@ -112,21 +112,29 @@
     (`duplicates t)
     (`no-cache t)   ;Not much can be done here, as long as we handle
-                    ;non-prefix matches.
+                                        ;non-prefix matches.
```
Unnecessary reindentation

joaotavora commented:
> Oh good, right? i.e. looking for face is also correct, right?

Actually, we look for `font-lock-face`.

> "similar to"

Fixed

> unnecessary reindentation

Fixed. When you can, fix those semicolons there, my `aggressive-indent-mode` loves to blow it up, lost count how many times I fixed this.

dgutov commented:
Thank you very much for your work :+1:
